raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,568 @@
+    {
+        uint participant1_amount;
+        uint participant2_amount;
+        uint total_deposit;
+
+        Participant memory participant1_state = channels[channel_identifier].participants[participant1];
+        Participant memory participant2_state = channels[channel_identifier].participants[participant2];
+
+        // Make sure the addresses are channel participant addresses
+        require(participant1_state.initialized);
+        require(participant2_state.initialized);
+
+        // Direct token transfers done through the token `transfer` function
+        // cannot be accounted for, these superfluous tokens will be burned,
+        // this is because there is no way to tell which participant (if any)
+        // had ownership over the token.
+        total_deposit = participant1_state.balance + participant2_state.balance;
+
+        participant1_amount = (
+            participant1_state.balance
+            + participant2_state.transferred_amount
+            - participant1_state.transferred_amount
+        );
+
+        // To account for cases when participant2 does not provide participant1's balance proof
+        // Therefore, participant1's transferred_amount will be lower than in reality
+        participant1_amount = min(participant1_amount, total_deposit);
+
+        // To account for cases when participant1 does not provide participant2's balance proof
+        // Therefore, participant2's transferred_amount will be lower than in reality
+        participant1_amount = max(participant1_amount, 0);
+
+        // At this point `participant1_amount` is between [0,total_deposit], so this is safe.
+        participant2_amount = total_deposit - participant1_amount;
+
+        // Remove the channel data from storage
+        delete channels[channel_identifier];
+        delete closing_requests[channel_identifier];
+
+        // Do the actual token transfers
+        require(token.transfer(participant1, participant1_amount));
+        require(token.transfer(participant2, participant2_amount));
+
+        ChannelSettled(channel_identifier);
+    }
+
+    // TODO
+    /*function cooperativeSettle(
+        uint channel_identifier,
+        uint256 balance1,
+        uint256 balance2,
+        bytes signature1,
+        bytes signature2)
+        public
+    {
+
+    }*/
+
+    /*
+     * Internal Functions
+     */
+
+    function updateParticipantStruct(
+        uint channel_identifier,
+        address participant,
+        uint64 nonce,
+        bytes32 locksroot,
+        uint256 transferred_amount)
+        internal
+    {
+        Channel storage channel = channels[channel_identifier];
+
+        require(channel.participants[participant].initialized);
+        require(nonce > channel.participants[participant].nonce);
+        // Transfers can have 0 value
+        require(transferred_amount >= channel.participants[participant].transferred_amount);
+        require(locksroot != channel.participants[participant].locksroot);
```

Comment:
This require is incorrect, the locksroot may be '0x0' and it may not change between two balance proofs:

- A balance proof, were there are no pending locks, will have the value "0x0", which is equal the unintialized value.
- A second call to update could be with a direct transfer balance proof, were the transferred_amount increased but the locksroot stayed the same.

Reply:
fixed https://github.com/raiden-network/raiden/pull/1286/commits/2bc0095f01937e6b7e854ade02a3ded45a4fba7a